File: typescript/src/context.ts
// =============================================================================
// Opper SDK - Trace Context (AsyncLocalStorage)
// =============================================================================

import { AsyncLocalStorage } from "node:async_hooks";

/** The current trace context, propagated automatically via AsyncLocalStorage. */
export interface TraceContext {
  readonly spanId: string;
  readonly traceId: string;
  /** Set by tool tracing wrappers so sub-agents skip redundant span creation. */
  readonly isToolSpan?: boolean;
}

const traceStore = new AsyncLocalStorage<TraceContext>();

/** Get the current trace context, if inside a `traced()` block. */
export function getTraceContext(): TraceContext | undefined {
  return traceStore.getStore();
}

/** Run a function with the given trace context. */
export function runWithTraceContext<T>(ctx: TraceContext, fn: () => T): T {
  return traceStore.run(ctx, fn);
}

File: python/src/opperai/_context.py
"""Opper SDK — Trace context propagation via contextvars."""

from __future__ import annotations

from contextvars import ContextVar
from dataclasses import dataclass


@dataclass(frozen=True)
class TraceContext:
    """Holds the current span ID and trace ID for automatic propagation."""

    span_id: str
    trace_id: str
    is_tool_span: bool = False


_trace_context: ContextVar[TraceContext | None] = ContextVar("_opper_trace_context", default=None)


def get_trace_context() -> TraceContext | None:
    """Get the current trace context, if any."""
    return _trace_context.get()


def set_trace_context(ctx: TraceContext | None) -> None:
    """Set the current trace context."""
    _trace_context.set(ctx)
